ops/testing.py

```diff
@@ -1759,6 +1789,104 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class _MockStorage:
+    def __init__(self, location, src):
+        self._src = src
+        os.makedirs(src, exist_ok=True)
+        self._location = location
+
+    def contains(self, path):
+        return str(path).startswith(self._location)
```

Comment:
This could also use a docstring. Because it defines a contains as in 'exists inside my directory'. But it doesn't mean that the path actually exists.
Also, I *think* there is the potential for it being wrong as:
```
'/foo/bar/baz'.startswith('/foo/b')
```
*If* we require that _location ends with a '/' then we should document it, and enforce it in `__init__`. (And, in fact, I would guess the default is just the location defined in the charm metadata, and 99% of the time does *not* end in a '/'.)

We could also just use `pathlib` and use things like:  `is_relative_to(other)`
https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to

IIRC pathlib is in python 3.5, but not all OS and System functions allow you to use it. (which is why a bunch of stuff uses `os.SOMETHING(str(path))`. But I would still recommend using pathlib for the Storage object.

Reply:
Indeed you are right on the potential bug.  Should be addressed now - I've updated the entire _TestingStorage class to use pathlib everywhere.